src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -0,0 +1,728 @@
+    ):
+        if dependent_columns == COLUMN_DEPENDENCY_ALL:
+            can_flatten = False
+        else:
+            for dc in dependent_columns:
+                dc_state = subquery_column_states.get(dc)
+                if dc_state and dc_state.change_state in (
+                    (
+                        ColumnChangeState.CHANGED_EXP,
+                        ColumnChangeState.DROPPED,
+                        ColumnChangeState.NEW,
+                    )
+                ):
+                    can_flatten = False
+                    break
+    return can_flatten
+
+
+def can_clause_dependent_columns_flatten(
+    dependent_columns: Optional[Set[str]], subquery_column_states: ColumnStateDict
+):
+    can_flatten = True
+    if dependent_columns == COLUMN_DEPENDENCY_DOLLAR:
+        can_flatten = False
+    elif (
+        subquery_column_states.has_changed_columns
+        or subquery_column_states.has_new_columns
+    ):
+        if dependent_columns == COLUMN_DEPENDENCY_ALL:
+            can_flatten = False
+        else:
+            for dc in dependent_columns:
+                dc_state = subquery_column_states.get(dc)
+                if dc_state:
+                    if dc_state.change_state == ColumnChangeState.CHANGED_EXP:
+                        can_flatten = False
+                        break
+                    elif dc_state.change_state == ColumnChangeState.NEW:
+                        # Most of the time this can be flatten. But if a new column uses window function and this column
+                        # is used in a clause, the sql doesn't work in Snowflake.
+                        # For instance `select a, rank() over(order by b) as d from test_table where d = 1` doesn't work.
+                        # But `select a, b as d from test_table where d = 1` works
+                        # We can inspect whether the referenced new column uses window function. Here we are being
+                        # conservative for now to not flatten the SQL.
+                        can_flatten = False
+                        break
+    return can_flatten
+
+
+def initiate_column_states(column_names: List[str]):
+    column_states = ColumnStateDict()
+    for i, name in enumerate(column_names):
+        column_states[name] = ColumnState(
+            name,
+            index=i,
+            change_state=ColumnChangeState.UNCHANGED_EXP,
+            expression=UnresolvedAttribute(quote_name(name)),
+            dependent_columns=COLUMN_DEPENDENCY_EMPTY,
+            depend_on_same_level=False,
+            referenced_by_same_level_columns=COLUMN_DEPENDENCY_EMPTY,
+            state_dict=column_states,
+        )
+    return column_states
+
+
+def derive_column_states_from_subquery(
+    cols: Iterable[Expression], from_: Selectable
+) -> ColumnStateDict:
+    analyzer = from_.analyzer
+    column_states = ColumnStateDict()
+    column_index = 0
+    # populate column status against subquery
```

Comment:
Instead of commenting about begin and end of code block, can we extract them into a function instead?